reference/test/matrix/dense_kernels.cpp

```diff
@@ -780,3061 +779,2103 @@ TYPED_TEST(Dense, MovesToPrecision)
+    using T = typename TestFixture::value_type;
+    using MixedT = typename MixedMtx::value_type;
+    auto alpha = gko::initialize<MixedMtx>({8.0}, this->exec);
 
-    auto ref_permuted =
-        gko::as<Mtx>(gko::as<Mtx>(mtx->inverse_row_permute(&permute_idxs))
-                         ->inverse_column_permute(&permute_idxs));
-    mtx->inverse_permute(&permute_idxs, permuted);
+    {
+        auto conversion = gko::make_temporary_conversion<T>(alpha);
+        conversion->at(0, 0) = T{7.0};
+    }
 
-    GKO_ASSERT_MTX_NEAR(permuted, ref_permuted, 0.0);
-    ASSERT_EQ(permuted->get_stride(), 4);
+    ASSERT_EQ(alpha->at(0, 0), MixedT{7.0});
 }
 
 
-TYPED_TEST(Dense, NonSquareMatrixInversePermuteIntoDenseFails)
+TYPED_TEST(Dense, MakeTemporaryConversionConstDoesntConvertBack)
 {
-    using Mtx = typename TestFixture::Mtx;
-    auto exec = this->mtx4->get_executor();
-    gko::array<gko::int32> permute_idxs{exec, {1, 2, 0}};
+    using MixedMtx = typename TestFixture::MixedMtx;
+    using T = typename TestFixture::value_type;
+    using MixedT = typename MixedMtx::value_type;
+    auto alpha = gko::initialize<MixedMtx>({8.0}, this->exec);
 
-    ASSERT_THROW(
-        this->mtx4->inverse_permute(&permute_idxs, this->mtx4->clone()),
-        gko::DimensionMismatch);
+    {
+        auto conversion = gko::make_temporary_conversion<T>(
+            static_cast<const MixedMtx*>(alpha.get()));
+        alpha->at(0, 0) = MixedT{7.0};
+    }
+
+    ASSERT_EQ(alpha->at(0, 0), MixedT{7.0});
 }
 
 
-TYPED_TEST(Dense,
-           SquareMatrixInversePermuteIntoDenseFailsForWrongPermutationSize)
+TYPED_TEST(Dense, ScaleAddIdentityRectangular)
 {
-    using Mtx = typename TestFixture::Mtx;
-    auto exec = this->mtx5->get_executor();
-    gko::array<gko::int32> permute_idxs{exec, {0, 1}};
+    using T = typename TestFixture::value_type;
+    using Vec = typename TestFixture::Mtx;
+    using MixedVec = typename TestFixture::MixedMtx;
+    auto alpha = gko::initialize<Vec>({2.0}, this->exec);
+    auto beta = gko::initialize<Vec>({-1.0}, this->exec);
+    auto b = gko::initialize<Vec>(
+        {I<T>{2.0, 0.0}, I<T>{1.0, 2.5}, I<T>{0.0, -4.0}}, this->exec);
 
-    ASSERT_THROW(
-        this->mtx5->inverse_permute(&permute_idxs, this->mtx5->clone()),
-        gko::ValueMismatch);
+    b->add_scaled_identity(alpha, beta);
+
+    GKO_ASSERT_MTX_NEAR(b, l({{0.0, 0.0}, {-1.0, -0.5}, {0.0, 4.0}}), 0.0);
 }
 
 
-TYPED_TEST(Dense, SquareMatrixInversePermuteIntoDenseFailsForWrongDimensions)
+template <typename ValueIndexType>
+class DenseWithIndexType
+    : public Dense<
+          typename std::tuple_element<0, decltype(ValueIndexType())>::type> {
+public:
+    using value_type =
+        typename std::tuple_element<0, decltype(ValueIndexType())>::type;
+    using index_type =
+        typename std::tuple_element<1, decltype(ValueIndexType())>::type;
+
+    index_type invalid_index = gko::invalid_index<index_type>();
```

Comment:
is that one even used? In the `assert_***_eq_mtx` functions the invalid index is defined again.